security-admin/src/main/java/org/apache/ranger/security/web/filter/AltiscaleRangerAuthFilter.java

```diff
@@ -0,0 +1,515 @@
+	 * to create a ranger session if the request is authenticated.
+	 */
+	@Override
+	protected void doFilter(FilterChain filterChain, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+		String userName = null;
+		boolean checkCookie = response.containsHeader("Set-Cookie");
+		if (checkCookie) {
+			Collection<String> authUserName = response.getHeaders("Set-Cookie");
+			if (authUserName != null) {
+				userName = extractUserNameFromCookieHeader(authUserName);
+			}
+		}
+		// if security context does not have a user session, authenticate the security context and create a session
+		createSessionForUser(userName, request);
+
+		// Delegate call to next filters
+		super.doFilter(filterChain, request, response);
+	}
+
+	/**
+	 *
+	 * @param userName
+	 * @param request
+	 */
+	private void createSessionForUser (String userName, HttpServletRequest request) {
+		if (!isAuthenticated()) {
+			String defaultUserRole = "ROLE_USER";
+			// if the userName is found on the token then log into ranger using the same user
+			if (userName != null && !userName.trim().isEmpty()) {
+				final List<GrantedAuthority> grantedAuths = new ArrayList<>();
+				grantedAuths.add(new SimpleGrantedAuthority(defaultUserRole));
+				final UserDetails principal = new User(userName, "", grantedAuths);
+				final Authentication finalAuthentication = new UsernamePasswordAuthenticationToken(principal, "", grantedAuths);
+				WebAuthenticationDetails webDetails = new WebAuthenticationDetails(request);
+				((AbstractAuthenticationToken) finalAuthentication).setDetails(webDetails);
+				RangerAuthenticationProvider authenticationProvider = new RangerAuthenticationProvider();
+				authenticationProvider.setAlt_ssoEnabled(true);
+				Authentication authentication = authenticationProvider.authenticate(finalAuthentication);
+				authentication = getGrantedAuthority(authentication);
+				SecurityContextHolder.getContext().setAuthentication(authentication);
+			}
+		}
+	}
+
+	/**
+	 *
+	 * @param authUserName
+	 * @return the userName
+	 */
+	private String extractUserNameFromCookieHeader (Collection<String> authUserName) {
+		String userName = null;
+		Iterator<String> i = authUserName.iterator();
+		while (i.hasNext()) {
+			String cookie = i.next();
+			if (!StringUtils.isEmpty(cookie)) {
+				if (cookie.toLowerCase().startsWith(AuthenticatedURL.AUTH_COOKIE.toLowerCase()) && cookie.contains("u=")) {
+					userName = extractUserNameFromCookie(cookie);
+				}
+			}
+		}
+		return userName;
+	}
+
+	/**
+	 *
+	 * @param cookieStr
+	 * @return the userName
+	 */
+	private String extractUserNameFromCookie (String cookieStr){
+		String userName = null;
+		if (cookieStr == null) {
+			return null;
+		}
+		String[] split = cookieStr.split(";");
+		if (split != null) {
+			for (String s : split) {
+				if (!StringUtils.isEmpty(s) && s.toLowerCase().startsWith(AuthenticatedURL.AUTH_COOKIE.toLowerCase())) {
```

Comment:
looks good to me, but these parsings are usually prone to errors, some of which can lead to security issues. Is there a library we can use that parses cookie and get username? If not, this is an excellent little method for which you can write a unittest and sleep better after putting enough examples.

Reply:
Right now, I am listing this as a TODO. Will check this later and try to use an existing library if possible. If not succeeded, I will write a small one myself.